### DISCUSSION TITLE
tutorials/voyage-part1/13-contributing-to-an-entity

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#74] [course feedback]
# tutorials/voyage-part1/13-contributing-to-an-entity

See how entities allow us to contribute fields across multiple subgraphs

https://www.apollographql.com/tutorials/voyage-part1/13-contributing-to-an-entity

**Comment by bizsystemanalyst** [comment]
Sorry i am not quiet following how in the arguments we are able to fetch the data source , I mean where was and how was  it(data sources) passed in the first place , that the resolver were abble to use the data sources .
Can you pls educate me on this ?

**Reply by mabuyo** [reply]
Hey there @bizsystemanalyst! We cover [how to connect and use data sources in Lift-off II](https://www.apollographql.com/tutorials/lift-off-part2/03-apollo-restdatasource). Hope that helps!

**Reply by torressam333** [reply]
Look at index.js of each subgraph and you will see datasources being passed in the context fn as a config option of `standAloneServer()`

```
 const { url } = await startStandaloneServer(server, {
            context: async () => {
                return {
                    dataSources: {
                        reviewsAPI: new ReviewsAPI()
                    }
                };
            },
            listen: { port }
        });
```

**Comment by MrChrisRodriguez** [comment]
Is there any guidance on how to mutate (create, edit, delete) a federated entity across subgraphs?

**Reply by mabuyo** [reply]
Hey @MrChrisRodriguez ! Trying to understand the scenario here. Taking an example in FlyBy, if we delete a location (from the `locations` subgraph domain), how might we also delete the reviews that are associated with it (that belongs in the `reviews` subgraph)?

**Reply by MrChrisRodriguez** [reply]
Yes, that's right.

I've seen a lot of explanations for how federation distributes queries, but haven't seen any explanations on how we should go about handling mutations in a "federation-friendly" way (i.e., what's best-practice?).

In my search, I've seen a lot of people asking the question, but the only resource I've found from Apollo is: [Federation is not a saga orchestrator](https://www.apollographql.com/docs/technotes/TN0040-federation-is-not-a-saga-orchestrator/)

In our specific use-case (with some simplification):
- We have a person-subgraph with a PersonInterface
- That subgraph defines several person-like entities that all implement the PersonInterface. One of these is Contact.
- We then have a crm-subgraph that adds a bunch of crm-specific fields and capabilities to the Contact.

In the CRM application, my customer will want to create new Contacts.
- How should I approach using the PersonInterface and the Contact's respective mutations?
- What if I'm bulk importing thousands or tens of thousands of Contacts?

I'd think that an entity implementing an interface would be able to easily call the interface's constructor but I just haven't seen a any clear-cut guidance on the matter and was hoping you should shine some light on it.

**Reply by MrChrisRodriguez** [reply]
@mabuyo just following up to see if you have any thoughts on the above!

**Reply by mabuyo** [reply]
Thank you so much for the ping @MrChrisRodriguez. This is outside my domain knowledge, but I reached out to a teammate and here's what they had to say!

> We don't currently have any other resources for mutations in federation since mutations are specific to the subgraph that defines them and that logic should be owned by the subgraph needing to do the updates. If that update needs to touch on multiple datasets, that's where that [technote "Federation is not a saga orchestrator"](https://www.apollographql.com/docs/technotes/TN0040-federation-is-not-a-saga-orchestrator/) comes in!
> 
> In your case, ideally, you should make mutations that specifically update Contacts and design mutations around that use-case. If that needs to happen in the CRM, great. If that means that the CRM has to also update the other subgraph, then you can review the technote.
> 
> You don't have to return an interface or even care about it when using a mutation and you can treat it like updating/making any other type.

**Comment by yamalameyooooo** [comment]
the example here has a infinite looping query 
```
query Location($locationId: ID!) {
  location(id: $locationId) {
    id
    name
    description
    photo
    overallRating
    reviewsForLocation {
      location {
        reviewsForLocation {
          location {
            reviewsForLocation {
              location {
```
Any way we can avoid a looping scenario?

**Reply by mabuyo** [reply]
Hey @yamalameyooooo - great question! We don't cover this in the course, but limiting query depth is one of the things we recommend enforcing to [secure your supergraph](https://www.apollographql.com/docs/technotes/TN0021-graph-security/#limit-query-depth). The Apollo Router has some config options to limit query depth (along with other things), which you can [read more about in the docs](https://www.apollographql.com/docs/router/configuration/operation-limits/). 

You can also use query complexity analysis to calculate the cost of running such a query. [This technote](https://www.apollographql.com/docs/technotes/TN0021-graph-security/#use-rate-limiting-as-needed) goes into more detail.

Hope that helps!

**Comment by sunnytoolsmiths** [comment]
I made some changes following the instruction in this step. However when I test the query in explorer. It gives me the following error. Can you please help me understand how to interpret the error messaging here? 

{
  "name": "TypeError",
  "message": "Failed to fetch",
  "stack": "TypeError: Failed to fetch\n    at https://studio-ui-deployments.apollographql.com/build/static/js/app.0e8871c27f4e3448b3c0.js:2:128207\n    at r (https://edge.fullstory.com/s/fs.js:4:15190)\n    at e.<computed>.e.<computed> (https://studio-ui-deployments.apollographql.com/build/static/js/app.0e8871c27f4e3448b3c0.js:1119:23633)\n    at https://studio-ui-deployments.apollographql.com/build/static/js/5388.2d8f5f57d79433e99213.js:2:13172\n    at Generator.next (<anonymous>)\n    at https://studio-ui-deployments.apollographql.com/build/static/js/5388.2d8f5f57d79433e99213.js:2:8654\n    at new Promise (<anonymous>)\n    at F (https://studio-ui-deployments.apollographql.com/build/static/js/5388.2d8f5f57d79433e99213.js:2:8399)\n    at q (https://studio-ui-deployments.apollographql.com/build/static/js/5388.2d8f5f57d79433e99213.js:2:12729)\n    at https://studio-ui-deployments.apollographql.com/build/static/js/ExplorerPage.39fab69799334668896f.js:345:7987\n    at Generator.next (<anonymous>)\n    at l (https://studio-ui-deployments.apollographql.com/build/static/js/ExplorerPage.39fab69799334668896f.js:318:17368)"
}

**Reply by mabuyo** [reply]
Hey there @sunnytoolsmiths - this usually happens when the subgraph servers or the router aren't running. Can you check that all three components are running?

**Reply by sunnytoolsmiths** [reply]
you are right! I forgot to check the router. Thank you for the fast response!

**Comment by thecouchcoder** [comment]
I am not understanding why we needed to define a reference resolver in the locations subgraph for the Locations entity, but we don't need a reference resolver for locations in the reviews subgraph.

**Reply by mabuyo** [reply]
Hey there @thecouchcoder - good question! Apollo Server provided a default reference resolver for us behind the scenes. We cover this a little more thoroughly in the optional section "Learn more: Adding a reference resolver" (you'll have to click to expand). If you haven't yet, please read through the section and let us know if you have any follow-up questions!

<img width="1511" alt="Screenshot 2024-08-09 at 9 43 19 AM" src="https://github.com/user-attachments/assets/a14c10f5-5c16-4233-bde4-fc37040bb933">

**Comment by Ashlok2003** [comment]
The course is fantastic and easy to learn. I would like to thank the Apollo team ✌️ for creating such great content.
However, are there any sample projects available that follow production-grade practices and manage federation ? 
Additionally, when building and deploying an application on cloud services like AWS, how is such an application deployed and managed ? 
For example, in a simple e-commerce platform with different microservices like user service, product service, etc., how is federation managed and deployed over cloud ?

**Reply by mabuyo** [reply]
Hey there @Ashlok2003, we're so glad you enjoyed the course! :) And thanks for letting us know!

You can check out this repo for a sample project that deploys a federated graph (it's an e-commerce example with 8 subgraphs!) using Kubernetes and AWS/GCP: https://github.com/apollosolutions/reference-architecture

Refer to this [docs article for an overview of how it all works together](https://www.apollographql.com/docs/graphos/enterprise/reference-architecture/).

If you have any further questions about that information, the [community forums](https://community.apollographql.com/) is a good resource. Hope that helps!

**Comment by scottwgibson** [comment]
Thank you so far, just owuld like to point out a minor inconsistency.

The example under "Querying data across Subgraphs" has the following query

query GetLocationDetails($locationId: ID!) {
  location(id: $locationId) {
    id
    ....

with the example response showing the result of a locations Query instead.

{
  "data": {
    "locations": [
      {
      ...


Appreciate the course so far!

**Reply by mabuyo** [reply]
Great catch @scottwgibson - we'll get that fixed!
